### ISSUE TITLE
analysis data: arglists that come from metadata are not necessarily edn friendly and not necessarily content preserving

### ISSUE THREAD

**Issue by lread** [clj-kondo/clj-kondo#2096]
## version

```
❯ clj-kondo --version
clj-kondo v2023.05.26
```
## Intro
Clj-kondo analysis data, when requested to do so, includes `:arglist-strs`. 
This string encoding of arglists nicely preserves the arglists as written in the code and is edn-friendly.

## Reference
[Discussion on Slack](https://clojurians.slack.com/archives/CHY97NXE2/p1685662451257919?thread_ts=1685661153.172639&cid=CHY97NXE2) with @borkdude and @ericdallo.

## Problem
When using clj-kondo's analysis data, `:arglist-strs` only shows the original arglists.
It does not incorporate `:arglists` specified as metadata. 

This is problematic for tooling that would like to show arglists as documentation as the author of the code intended.  Tooling can always refer to var analysis `:meta` `:arglists`, but this variant is not stringified and therefore does not always preserve arglists as written by the author, and is not necessarily edn-friendly.

## Repro
No `:arglists` metadata, all is good: `(defn foo [orig-arg])` -> `:arglist-strs ["[orig-arg]"]`
```Clojure
❯ clj-kondo --config '{:config-paths ^:replace [] :output {:format :edn :analysis {:var-definitions {:meta true} :arglists true}}}' \
        --lint - <<< $'(defn foo [orig-arg])' \
        | jet --query ':analysis :var-definitions first' --pretty
{:arglist-strs ["[orig-arg]"],
 :col 1,
 :defined-by clojure.core/defn,
 :defined-by->lint-as clojure.core/defn,
 :end-col 22,
 :end-row 1,
 :filename "<stdin>",
 :fixed-arities #{1},
 :meta nil,
 :name foo,
 :name-col 7,
 :name-end-col 10,
 :name-end-row 1,
 :name-row 1,
 :ns user,
 :row 1}
```

But for `(defn foo {:arglists '([new-arg])} [orig-arg])` -> `:arglist-strs ["[orig-arg]"]`.
We do see `:meta {:arglists (quote ([new-arg]))}` but it is not applied to `:arglist-strs`:
```Clojure
❯ clj-kondo --config '{:config-paths ^:replace [] :output {:format :edn :analysis {:var-definitions {:meta true} :arglists true}}}' \
        --lint - <<< $'(defn foo {:arglists \'([new-arg])} [orig-arg])' \
        | jet --query ':analysis :var-definitions first' --pretty
{:arglist-strs ["[orig-arg]"],
 :col 1,
 :defined-by clojure.core/defn,
 :defined-by->lint-as clojure.core/defn,
 :end-col 47,
 :end-row 1,
 :filename "<stdin>",
 :fixed-arities #{1},
 :meta {:arglists (quote ([new-arg]))},
 :name foo,
 :name-col 7,
 :name-end-col 10,
 :name-end-row 1,
 :name-row 1,
 :ns user,
 :row 1}
```

## Expected Behaviour
`:arglist-strs` should always reflect arglists even when overridden by metadata.
We can leave var analysis `:meta` `:arglists` as is.
Because `:arglists` metatada cannot be trusted, analysis data arities values should remain as is.

## Assumption
When overridden by metadata, no analysis data users are interested in original arglists.

#### option a) Always wrap in vector. 
Given `:arglists :fubar` we could encode as `:arglist-strs [":fubar"]`.
This also handles the case where user forgets the wrapping list, ex `:arglists [x y z]` becomes `:arglist-strs ["[x y z]"]`.
#### choice
option a seems like a better choice to me.

### 2) Handling of `^{arglists ...}`
Clojure does not seem to apply `^{:arglists ...}` for a `defn`.
```Clojure
user=> (defn ^{:arglists '([new])} foo [x])
#'user/foo
user=> (meta #'foo)
{:arglists ([x]), :line 1, :column 1, :file "NO_SOURCE_PATH", :name foo, :ns #object[clojure.lang.Namespace 0xbcb09a6 "user"]}
```

But it does do so for a `def`:
```Clojure
user=> (def ^{:arglists '([new])} foo)
#'user/foo
user=> (meta #'foo)
{:arglists ([new]), :line 1, :column 1, :file "NO_SOURCE_PATH", :name foo, :ns #object[clojure.lang.Namespace 0x42deb43a "user"]}
```

#### choice
Uh. Dunno. Probably doesn't matter too much? a?

### 3) Multiple `:arglists` are specified
This would be unusual but we should probably behave like Clojure does.

We could potentially have:
```Clojure
(defn ^{:arglists '([x1])} foo {:arglists '([x2])} ([x]) {:arglists '([x3])})
```
(but see nuance 2 for `^{:arglists ...}`).
As we learned for `:doc`, Clojure applies meta from left to right. So we should do the same for `:arglists`.

#### choice
no options here, just a nuance 

## Next steps 
We need to hash out nuances (and any other feedback), after that, happy to do up a PR if that is helpful.

**Comment by lread** [comment]
Hmmm... yes... interesting take.
Folks need to quote `:arglists` value because symbol resolution would otherwise fail.
```Clojure
user=> (defn foo {:arglists ([new])} [x])
Syntax error compiling at (REPL:1:22).
Unable to resolve symbol: new in this context
```
But how common is it that `:arglists` value requires more evaluation than just un`quote`ing?

**Comment by borkdude** [comment]
Note that with defn the way to override is:

```
user=> (defn foo "docstring" {:arglists '([{:keys [x y z]}])} [opts] (println opts))
#'user/foo
user=> (:arglists #'foo)
nil
user=> (:arglists (meta #'foo))
([{:keys [x y z]}])
```

Probably you knew this already, `^{:arglists ...}` is an edge case that we can leave behind.

I guess one column for the informal decision matrix here is that we can take the arglists from meta only if it's a single quoted value and ignore it otherwise.

**Comment by lread** [comment]
> Note that with defn the way to override is:
> 
> ```
> user=> (defn foo "docstring" {:arglists '([{:keys [x y z]}])} [opts] (println opts))
> #'user/foo
> user=> (:arglists #'foo)
> nil
> user=> (:arglists (meta #'foo))
> ([{:keys [x y z]}])
> ```

Yup, was trying to get across that the reason folks quote arglists is that they have to, and not (probably typically anyway) because they need to have something evaluated.

> Probably you knew this already, `^{:arglists ...}` is an edge case that we can leave behind.

That's fine. Probably nuance 2 option b?

> I guess one column for the informal decision matrix here is that we can take the arglists from meta only if it's a single quoted value and ignore it otherwise.

That would work and is probably nuance 1 option b?

**Comment by borkdude** [comment]
Yes on both

**Comment by lread** [comment]
Perhaps worth sleeping on this one before proceeding?
And waiting to learn if @ericdallo can see any issues with this proposal?

**Comment by lread** [comment]
Funny observation: `:arglists` need to be evaluated because users typically don't want `:arglists` to be evaluated.

**Comment by borkdude** [comment]
> and therefore does not always preserve arglists as written by the author, and is not necessarily edn-friendly

I guess this is the crux.

**Comment by lread** [comment]
> > and therefore does not always preserve arglists as written by the author, and is not necessarily edn-friendly
> 
> I guess this is the crux.

For my use case, this is the crux.

**Comment by lread** [comment]
But maybe I could get around it via some analysis data `:hook`? Dunno have not explored these hooks yet.

**Comment by ericdallo** [comment]
both work for LSP side, just sharing [this piece of code](https://github.com/clojure-lsp/clojure-lsp/blob/15ba575a136fe52e9a259801a77e1d138891a32a/lib/src/clojure_lsp/kondo.clj#L97-L103) showing that we always try to edn/read-string if possible

**Comment by lread** [comment]
Ah interesting @ericdallo, thanks for sharing. I like the `:arglist-strs` string encoding because it avoids edn read failure cases that would happen for arglists like:
- a regex as a default: https://github.com/jaycfields/jry/blob/8307493e87617866357f3475108e80701ccbf8c8/src/clojure/jry/io.clj#L9
- an auto-resolved kw in destructuring: https://github.com/penpot/penpot/blob/a89d47b5c5c78cca256aecb4fb8c15623ea773d9/backend/src/app/rpc/commands/files.clj#L352

I was imagining edn friendliness was the reason clj-kondo has `:arglist-strs` and instead of an `:arglists` forms.

Cljdoc doesn't currently have a need to dig into the `:arglist-strs`, it just presents them (formatted with zprint). But I expect clojure-lsp needs the inner details and that is why it is converting `:arglist-strs` to forms.

**Comment by ericdallo** [comment]
Correct @lread, I agree with that encoding as well

**Comment by lread** [comment]
Thanks @ericdallo! So are you saying for your use cases proceeding with this issue makes sense? Note that we are currently on:
1. nuance 1 option b - Do not populate `:arglists-strs` when it does not conform
2. nuance 2 - We don't really care about this edge case

**Comment by ericdallo** [comment]
Yeah, I think it's ok, 1. could cause issues where user expects to see the arglists they typed even if wrong, but I think that's expected since they are really wrong.

**Comment by lread** [comment]
> Yeah, I think it's ok, 1. could cause issues where user expects to see the arglists they typed even if wrong, but I think that's expected since they are really wrong.

I too like nuance 1 option a because it would show what the author typed.
But maybe it would be helpful to also indicate arglists did not look like arglists.
I guess as consumers of the analysis data, we could always check `:meta` `:arglists` to see if they look valid and then optionally treat `:arglist-strs` differently. 

Theoretically, an author might dynamically generate `:arglists` within a `defn` declaration, but my grep.app skill did not find that happening in practice.

Fwiw, here's an example of current [cljdoc treatment of invalid-looking arglists](https://cljdoc.org/d/babbage/babbage/0.1.0/api/babbage.graph#defgraphfn) (click through to source and you'll see the author forgot wrapping vector).

So yeah, I think I still like nuance 1 option a. @borkdude do you think this would be problematic/a-bad-idea?

**Comment by borkdude** [comment]
Can you summarize what your proposal now is, with example, as short as possible?

**Comment by lread** [comment]
The text of this original issue remains valid and should be reviewed for details.

## To recap highlights
Currently `:arglist-strs` is not affected by `:arglists` metadata in analyzed code.
The benefits of `arglist-strs` encoding are:
1. it always preserves the original analyzed code
2. it is edn-friendly

Change:
1. Always apply `:arglists` metadata in analyzed code to `:arglist-strs`
2. Conversion to `:arglist-strs` will automatically un-quote `:arglists`
3. If an `:arglists` does not look like an arglists it will be converted to `["unrecognized arglists here"]`

## Example 1
```Clojure
(defn sh {:arglists '([cmd & args] [{:keys [env dir quiet?]} cmd & args])} [& args])
```
results in:
```Clojure
:arglist-strs ["[cmd & args]" "[{:keys [env dir quiet?]} cmd & args]"]
```
## Example 2
```Clojure
(defn oops {:arglists '(does not conform)} [a b c])
```
results in:
```Clojure
:arglist-strs ["does not conform"]
```

**Comment by borkdude** [comment]
Example 1 isn't a good example since you can't have more than 1 variadic overload, but other than that, it looks good.

I would say for example 2, it's better to fall back on the raw arglists-str that we had before, rather than inserting "does not conform".

More generally, I think it's better to _only_ change `:arglists-str` if it's an improvement and only if there is a single quoted list in the metadata.

**Comment by borkdude** [comment]
This thread is getting so long that I think a decision matrix would make things clearer. Feel free to start one on Google docs and invite me and Eric as the editors.

**Comment by lread** [comment]
Yes and communication can be tough! I thought you were asking me to recap the whole issue and now realize you were probably asking me to recap nuance 1 option a. We might be going in circles, a bit.

Anyhoo. 

Example 1 is from the wild:
https://github.com/metabase/metabase/blob/3384ecbb4c5cb2db6f1e8ee4d7e1130c40df1071/bin/build/src/metabuild_common/shell.clj#L73-L74
I guess they are just trying to document different ways to call the fn.

Example 2
To be clear, the proposal is to populate invalid looking arglists in `:arglist-strs` as the author wrote them and not some static string. Even if the arglists look invalid it would be nice to capture them so that we can choose to display them.

I can do up a grid.

**Comment by lread** [comment]
The decision matrix is powerful my friends, it does help one to focus on the crux.

Here's my first [stab](https://docs.google.com/spreadsheets/d/1V_TvfEGJgu4AluH2IQ37IhV9mF9LRSysOKZxrRMX90Y/edit?usp=sharing).

I've invited @borkdude and @ericdallo as editors.
If others have an interest in contributing to how this issue gets addressed, lemme know and I'll add you as well.

**Comment by borkdude** [comment]
@lread I read decision 1 and I'm fine with choosing the proposed candidate.

For decision 2: I have trouble imagining what all the words mean. Could you please add an example arglists and update what the alternatives for decision 2 would look like (in the matrix, not here)?

Small note: the colors should be used to emphasize cells, not column or row headers, so you can squint and see that one column is "mostly green".

**Comment by lread** [comment]
Thanks for taking a peek!

> For decision 2: I have trouble imagining what all the words mean. Could you please add an example arglists and update what the alternatives for decision 2 would look like (in the matrix, not here)?

Yeah, can do. The crux for suspicious arglists, if we decide they should not be omitted, is that we'd also want an encoding that is edn-friendly and content preserving.

> Small note: the colors should be used to emphasize cells, not column or row headers, so you can squint and see that one column is "mostly green".

Right you are. Thanks.

**Comment by ericdallo** [comment]
Thanks for the decision matrix @lread, it really helps!
I really like the decision 1 approach 2

**Comment by lread** [comment]
Ok, very good, it sounds like we have consensus on decision 1! 🎉 

That leaves us with only decision 2 to deal with.

**Comment by borkdude** [comment]
Agreed on 1.2

My proposal on 2 => only do this when arglist looks like `'([...] [...])`, else do not override (and preserve whatever was there before).
So only if the users single-quoted a list of vectors, we will override the `:arglists-str`. This seems to be the most common case and an improvement compared to the status quo. All other cases, we will ignore, until we learn more.
